### PR TITLE
overlord/snapstate: allow core defaults configuration via 'system' key

### PR DESCRIPTION
**Pull request by bboozzoo**
Resurrected #4597 and updated to use the 'system' key as agreed in https://forum.snapcraft.io/t/development-sprint-march-5th-2018/4345/15

Had to update some tests in `overlord/configstate` as code assumes now that the 'core' snap is installed.

### CODE REVIEW

File: overlord/snapstate/snapstate.go [hunk 1]
```diff
@@ -1761,8 +1763,16 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
...
+	core, err := CoreInfo(st)
+	if err != nil {
+		return nil, err
+	}
+	isCoreDefaults := core.Name() == snapName
+
 	si := snapst.CurrentSideInfo()
-	if si.SnapID == "" {
+	// core snaps can be addressed even without a snap-id via the special
+	// "system" value in the config
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: I had to chase how this works,  probably worth adding

// first-boot always configures the core snap with UseConfigDefaults

Basically we don't need to worry  about:

https://github.com/bboozzoo/snapd/blob/e1ebc92eb7355c1202ad42ed071c385aaa280d14/overlord/snapstate/snapstate.go#L275

File: overlord/snapstate/snapstate.go [hunk 2]
```diff
@@ -1771,6 +1781,17 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
 		return nil, err
 	}
 
+	// we support setting core defaults via "system"
+	if isCoreDefaults {
+		if defaults, ok := gadgetInfo.Defaults["system"]; ok {
+			if _, ok := gadgetInfo.Defaults[si.SnapID]; ok && si.SnapID != "" {
+				return nil, fmt.Errorf("cannot have both system and core-snap-id defaults")
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go]: Will add

File: overlord/snapstate/snapstate.go [hunk 3]
```diff
@@ -1761,8 +1763,16 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
...
+	if err != nil {
+		return nil, err
+	}
+	isCoreDefaults := core.Name() == snapName
+
 	si := snapst.CurrentSideInfo()
-	if si.SnapID == "" {
+	// core snaps can be addressed even without a snap-id via the special
+	// "system" value in the config
+	if si.SnapID == "" && !isCoreDefaults {
```
[the remark this answers is not in the document]
Comment (bboozzoo) [reply overlord/snapstate/snapstate.go]: Will add one

File: overlord/snapstate/snapstate_test.go [hunk 4]
```diff
@@ -7619,6 +7622,63 @@ func (s *snapmgrTestSuite) TestConfigDefaults(c *C) {
...
+`)
+
+	makeInstalledMockCoreSnap(c)
+
+	defls, err := snapstate.ConfigDefaults(s.state, "core")
+	c.Assert(err, IsNil)
+	c.Assert(defls, DeepEquals, map[string]interface{}{"foo": "bar"})
+}
+
+func (s *snapmgrTestSuite) TestConfigDefaultsSystemConflictsCoreSnapId(c *C) {
```
Comment (stolowski) [inline overlord/snapstate/snapstate_test.go]: Thank you

File: overlord/snapstate/snapstate.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (pedronis) [reply overlord/snapstate/snapstate.go]: I wonder if people will need to write gadgets that support both for a bit, in which case logging and preferring system would work maybe better

**Comment by bboozzoo** [comment]
@pedronis do you suggest doing something as simple as:
```
diff --git a/overlord/snapstate/snapstate.go b/overlord/snapstate/snapstate.go
index da977bdfd..4d1219113 100644
--- a/overlord/snapstate/snapstate.go
+++ b/overlord/snapstate/snapstate.go
@@ -1874,7 +1874,7 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
        if isCoreDefaults {
                if defaults, ok := gadgetInfo.Defaults["system"]; ok {
                        if _, ok := gadgetInfo.Defaults[si.SnapID]; ok && si.SnapID != "" {
-                               return nil, fmt.Errorf("cannot have both system and core-snap-id defaults")
+                               logger.Noticef("core snap configuration defaults found under both 'system' key and core-snap-id")
                        }
 
                        return defaults, nil
```
or merging both maps with one under `system` being preferred?

File: overlord/snapstate/snapstate.go [hunk 5]
```diff
@@ -1874,7 +1874,7 @@ func ConfigDefaults(st *state.State, snapName string) (map[string]interface{}, e
 	if isCoreDefaults {
 		if defaults, ok := gadgetInfo.Defaults["system"]; ok {
 			if _, ok := gadgetInfo.Defaults[si.SnapID]; ok && si.SnapID != "" {
-				return nil, fmt.Errorf("cannot have both system and core-snap-id defaults")
+				logger.Noticef("core snap configuration defaults found under both 'system' key and core-snap-id")
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go]: it should probably say, "... core snap-id, preferring 'system'"

**Pull request merged**